### DISCUSSION TITLE
65. 트랜잭션 격리 수준에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#67] [General]
- MySQL InnoDB의 기본 트랜잭션 격리 수준은 무엇인가요?
- MySQL InnoDB에서 굳이 SERIALIZABLE 격리 수준을 기본으로 사용하지 않는 이유는 무엇인가요?

**Comment by eun-byeol** [comment]
트랜잭션 격리 수준은 여러 트랜잭션이 동시에 처리될 때 특정 트랜잭션이 다른 트랜잭션에서 변경하거나 조회하는 `데이터를 볼 수 있게 허용할지 말지 결정하는 것`입니다.

`READ UNCOMMITTED`, `READ COMMITTED`, `REPEATABLE READ`, `SERIALIZABLE`로 4가지로 나뉩니다. 뒤로 갈수록 고립 정도가 높아지지만 동시 처리 성능은 떨어집니다.

MySQL InnoDB 엔진은 기본적으로 `REPEATABLE READ` 격리 수준을 사용합니다. InnoDB의 경우 REPEATABLE READ 수준에서 `Next-Key Lock`으로 `Phantom Read`를 방지할 수 있기 때문입니다.

**Comment by eun-byeol** [comment]
## 트랜잭션 격리 수준
- READ UNCOMMITTED : 커밋되지 않는 읽기
- READ COMMITTED : 커밋된 읽기
- REPEATABLE READ : 반복 가능한 읽기
- SERIALIZABLE : 직렬화 가능

  | Dirty Read | Non-Repeatable Read | Phantom Read
-- | -- | -- | --
READ UNCOMMITTED | O | O | O
READ COMMITTED | X | O | O
REPEATABLE READ | X | X | O(InnoDB는 없음)
SERIALIZABLE | X | X | X


## 부정합
- `Dirty Read` : 트랜잭션 작업이 완료되지 않았는데 다른 트랜잭션에서 볼 수 있는 현상
- `Non-Repeatable Read` : 하나의 트랜잭션 내에서 똑같은 select 쿼리 실행 시 같은 결과를 가져오지 못하는 현상
- `Phantom Read` : 다른 트랜잭션에서 수행한 변경 작업에 의해 레코드가 보였다 안 보였다 하는 현상

## InnoDB가 Phantom Read를 방지하는 방법
잠금을 동반한 읽기의 경우(SELECT … FOR UPDATE, SELECT … FOR SHARE) 언두 레코드에 잠금을 걸 수 없어, 현재 레코드의 값을 가져온다. 이 경우 InnoDB의 Next-Key Lock이 활성화된다. 기존 레코드(레코드 락)와 레코드 사이의 간격(갭 락)을 잠가 새로운 레코드 추가를 방지하여 팬텀 리드를 방지한다.

**Reply by jongmee** [reply]
InnoDB가 Phantom Read를 방지하는 방법에서 Next-Key Lock이 활성화 되는 경우를 약간 더 보충할게요 ☺️

1. select for update 혹은 select .. lock in share mode (명시적 잠금 쿼리)
2. 인덱스 범위 검색 
    
    `SELECT * FROM products WHERE price BETWEEN 100 AND 200 FOR UPDATE`
    
3. **정확히 하나의 레코드**만을 지정하지 않는 쿼리 (UNIQUE/PRIMARY KEY 조건에 해당하지 않는 경우)
4. 인덱스가 없는 경우 **테이블 전체를 잠금**하는 **Gap Lock**이 발생할 수 있다.
    
    `SELECT * FROM orders WHERE status = 'pending' FOR UPDATE` 
    
5. AUTO_INCREMENT 열에 대해 Insert 시

생각보다 엄청 많아서 데드락 발생의 주원인이라고 합니다 🥲 (이 발언은 당근 테크 블로그 DBA의 글에서 봤었음)
